### ISSUE TITLE
Litespeed caching voorkomt dat site geladen wordt

### ISSUE THREAD

**Issue by simonbesters** [rudiedirkx/wappalyzer#20]
Er zijn websites die [Litespeed](https://litespeedtech.com/) gebruiken als webserver. Deze sites worden niet goed geladen in Wappalyzer:

https://app.digitalinsightsplatform.nl/checks/3197454
https://app.digitalinsightsplatform.nl/checks/3197457
https://app.digitalinsightsplatform.nl/checks/3197459

Door de volgende no-cache headers mee te sturen zorgen we dat deze pagina's toch wel goed geladen worden:

```
 'Cache-Control': 'no-cache',
 'Pragma': 'no-cache',
```

Het werkt alleen als tevens dit deel van `async open` wordt overgeslagen (fake blank page laden):

```javascript
      await page.setRequestInterception(true)

      page.on('request', (request) =>
        request.respond({
          status: 200,
          contentType: 'text/plain',
          body: 'ok',
        })
      )
```

Waarom doet Wappa dit eigenlijk?

**Comment by rudiedirkx** [comment]
Als ik m lokaal draai met headers via cli, vindt ie nog steeds geen resultaten:

    node node_modules/wappalyzer/cli.js 'https://www.werkenbijutrecht.nl/' --pretty -e -H 'Pragma: no-cache' -H 'Cache-Control: no-cache'

Zo deed jij dat toch ook? Dus de Wappa code zoals ie op Prod is, en met `wappalyzer-custom-technologies.json`, en met 2 extra headers.

**Comment by simonbesters** [comment]
Ik had het in de driver gedaan:
vanaf line 461:
```javascript
      await page.setRequestInterception(true)

		page.on('request', (request) => {
			const customHeaders = {
				'Cache-Control': 'no-cache',
				'Pragma': 'no-cache',
			};

			request.continue({
				headers: {
					...request.headers(),
					...customHeaders
				}
			});
		});


      await page.goto(url)
```

Maar ik weet nog steeds niet goed waarom die fake response daar gebruikt wordt.

**Comment by rudiedirkx** [comment]
Aw man hoe werkt Wappalyzer ook alweer. Ik heb het helemaal uitgevogeld toen we forkten, maar daarna weer vergeten. Hij doet 2 rondjes requests volgens mij. `open()` is de eerste, en daar doet ie niks met de response. Die request doet ie alleen om localStorage en sessionStorage te setten. Die `if (true) {` van een paar regels erboven was vroeger een `if I should set *storage`, maar dat was altijd true, dus dat heeft nooit gewerkt zoals intended (DENK IK). Die eerste request doet ie dus altijd, ook al heeft ie geen storage om te setten. Maar daarna doet ie pas de echte requests waar ie op headers en JS etc gaat letten. Raar dat no-cache headers _daar_ werken. Ik zou denken dat ze in `goto()` moeten. Misschien omdat de eerste request (`open()`) zo niet gecacht wordt, en ook al wordt de 2e (`goto()`) wel gecacht, daar zou de 3e request pas last van hebben en die doen wij nooit. Misschien werkt het zo. Maar het moet eigenlijk ook wel werken voor 3 requests, of `recursive`.

**Comment by rudiedirkx** [comment]
`request.continue()` met headers vertelt de client dat ie de response niet moet cachen, niet de server dat ie de request/response niet moet cachen... Als wij nooit cache willen (dat denk ik) kunnen we dat ook anders zeggen. Dan is het alleen client-side cache, wie weet wat de server/website zelf wel cacht. Als jouw oplossing werkt lijkt server cache geen probleem te zijn.

    await page.setCacheEnabled(false);

disablet alle client-side cache misschien.

**Comment by rudiedirkx** [comment]
> `request.continue()` met headers vertelt de client dat ie de response niet moet cachen, niet de server dat ie de request/response niet moet cachen...

Nee, is niet waar. Wel naar de server, en daarom werkt die `respond()` niet meer. Maar dat wil ik niet veranderen. We moeten beter cache headers op kunnen sturen.

**Comment by rudiedirkx** [comment]
Nee, ik snap het niet. CLI `-H` stuurt de headers door met alle `goto()` requests (wat ik zou willen), en die zouden er juist toe moeten doen. Als ik die initial request van `open()` niet eens doe (dus `if (false)` maken) dan werkt het nooit, met welke headers dan ook. Dan vindt ie nooit front-end resultaten. Waarom zou alleen in de initial request no-cache opsturen werken??

**Comment by simonbesters** [comment]
Nou hij doet toch niks met de headers (althans niet voor Litespeed):

Als ik alleen dit over laat, dan doet hij het dus wel met Litespeed:      

```javascript
await page.setRequestInterception(true)

      page.on('request', (request) => {
		request.continue()}
      )

      await page.goto(url)
```

**Comment by simonbesters** [comment]
Je kunt je zelfs afvragen waarom die hele stap er tussen moet blijven. Ik weet niet wat het doet en waarom het die dingen doet.

**Comment by rudiedirkx** [comment]
Die `continue()` doet vaker iets onverklaarbaars en ik snap m nooit. Die zijn we al een paar keer tegengekomen. Ook bij Cookies.

Is dat de `continue()` van de initiele `open()`? Niet eens bij 'alle' `goto()` page loads (wij doen er maar 1)? Dan is het nog raarder dat het iets doet. Dat is helemaal geen belangrijke request... _Daarna_ begint het verzamelen van info pas.

**Comment by rudiedirkx** [comment]
In `goto()` doet ie ook altijd een `continue()` (of `abort()`). We hebben gewoon te weinig `continue()`s. Logisch... 👀 

Ik ga die `request.continue()` toevoegen in de `open()` request op test1 en test3 workers en dan vanavond kijken wat er gebeurt. Ik heb geen idee waarom het iets doet, en nog minder idee waarom het _beter_ werkt (in ieder geval met Litespeed), mja, als het maar minstens net zo goed werkt.

**Comment by rudiedirkx** [comment]
Dit is de superleesbare uitstekende beste query om de resultaten van de autochecks van gisteren te vergelijken met de vorige keer. Nu even alleen voor WhatCms checks: https://app.digitalinsightsplatform.nl/custom-queries/141

**Comment by rudiedirkx** [comment]
We zitten trouwens wel in de publieke Wappalyzer repo over heel specifieke DIP dingen te praten...

**Comment by rudiedirkx** [comment]
Heel misschien relevant: https://github.com/peterbe/minimalcss/issues/223
